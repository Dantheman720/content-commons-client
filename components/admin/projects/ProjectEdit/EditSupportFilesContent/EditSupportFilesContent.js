/**
 *
 * EditSupportFilesContent
 *
 */
import React from 'react';
import { func, object, string } from 'prop-types';
import gql from 'graphql-tag';
import { graphql } from 'react-apollo';
import { Button, Form, Table } from 'semantic-ui-react';

import ModalItem from 'components/modals/ModalItem/ModalItem';
import VisuallyHidden from 'components/admin/projects/shared/VisuallyHidden/VisuallyHidden';
import EditSupportFileRow from 'components/admin/projects/ProjectEdit/EditSupportFileRow/EditSupportFileRow';

import { compareValues, capitalizeFirst } from 'lib/utils';

import './EditSupportFilesContent.scss';

/* eslint-disable react/prefer-stateless-function */
class EditSupportFilesContent extends React.PureComponent {
  getFileExtension = str => (
    str.slice( ( Math.max( 0, str.lastIndexOf( '.' ) ) || Infinity ) )
  )

  getFileExtensions = arr => {
    const allFileExtensions = arr.reduce( ( acc, curr ) => (
      acc.concat( this.getFileExtension( curr.filename ) )
    ), [] );
    const uniqueExtensions = [...new Set( allFileExtensions )];
    return uniqueExtensions;
  }

  getFilesByType = ( files, type ) => {
    /**
     * @todo Would be better to get file types from
     * a query. Need to figure out why these nested
     * arguments don't work:
     *
       query EditSupportFiles($id: ID!) {
        project: videoProject(id: $id) {
          srt: supportFiles(where: {filetype: "srt"}, orderBy: filename_ASC) {
            id
            filetype
            filename
          }
          other: supportFiles(where: {filetype_not: "srt"}, , orderBy: filetype_ASC) {
            id
            filetype
            filename
          }
        }
      }
     */
    if ( type === 'srt' ) {
      return files
        .filter( file => file.filetype === type )
        .sort( compareValues( 'filename' ) );
    }
    return files
      .filter( file => file.filetype !== 'srt' )
      .sort( compareValues( 'filetype' ) );
  }

  handleCancelClose = () => {
    console.log( 'cancel' );
    this.props.closeEditModal();
  }

  handleAddFiles = () => {
    console.log( 'add files' );
    this.addFilesInputRef.click();
  }

  handleAddFilesRef = input => {
    this.addFilesInputRef = input;
  }

  renderRow = file => {
    const { id } = file;
    const {
      fileType,
      projectId,
      data: { project: { supportFiles } }
    } = this.props;

    /**
     * see @todo comment in the getFilesByType method
     */
    const files = this.getFilesByType( supportFiles, fileType );

    return (
      <EditSupportFileRow
        key={ id }
        file={ file }
        projectId={ projectId }
        fileExtensions={ this.getFileExtensions( files ) }
      />
    );
  }

  render() {
    if ( !this.props.data.project ) return null;

    const {
      fileType,
      data: {
        error,
        loading,
        project: { supportFiles }
      }
    } = this.props;

    /**
     * see @todo comment in the getFilesByType method
     */
    const files = this.getFilesByType( supportFiles, fileType );

    if ( loading ) return 'Loading the support files...';
    if ( error ) return `Error! ${error.message}`;
    if ( !files || !files.length ) return null;

    const headline = fileType === 'srt'
      ? fileType.toUpperCase()
      : capitalizeFirst( fileType );

    return (
      <ModalItem
        className={ `edit-support-files ${fileType}` }
        headline={ `Edit ${headline} file${files.length > 1 ? 's' : ''} in this project` }
        textDirection="ltr"
      >
        <Form>
          <Table basic="very">
            <Table.Header>
              <Table.Row>
                <Table.HeaderCell width={ 9 }>
                  { `${headline} File${files.length > 1 ? 's' : ''} Selected` }
                </Table.HeaderCell>
                <Table.HeaderCell width={ 4 }>
                  Language
                  <small className="msg--required"> *</small>
                </Table.HeaderCell>
                <Table.HeaderCell />
              </Table.Row>
            </Table.Header>

            <Table.Body>
              { files.map( this.renderRow ) }
            </Table.Body>
          </Table>

          <div className="btn-group">
            <Button
              className="cancel-close"
              content="Cancel"
              basic
              size="tiny"
              onClick={ this.handleCancelClose }
              type="button"
            />
            <Button
              className="add-files"
              content="Add Files"
              color="blue"
              basic
              size="tiny"
              onClick={ this.handleAddFiles }
              type="button"
            />
            <VisuallyHidden>
              { /* eslint-disable jsx-a11y/label-has-for */
                /* eslint-disable jsx-a11y/label-has-associated-control */ }
              <label htmlFor="upload-file--multiple">upload files</label>
              <input
                id="upload-file--multiple"
                ref={ this.handleAddFilesRef }
                type="file"
                multiple
                tabIndex={ -1 }
              />
            </VisuallyHidden>
          </div>
        </Form>
      </ModalItem>
    );
  }
}

EditSupportFilesContent.propTypes = {
  data: object.isRequired,
  projectId: string.isRequired,
  fileType: string,
  closeEditModal: func
};

const SUPPORT_FILES_QUERY = gql`
  query SupportFiles($id: ID!) {
    project: videoProject(id: $id) {
      supportFiles {
        id
        filename
        filetype
        language {
          id
          displayName
        }
      }
    }
  }
`;

export default graphql( SUPPORT_FILES_QUERY, {
  options: props => ( {
    variables: {
      id: props.projectId
    },
  } )
} )( EditSupportFilesContent );
export { SUPPORT_FILES_QUERY };
