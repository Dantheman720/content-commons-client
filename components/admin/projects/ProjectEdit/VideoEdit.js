import React from 'react';

const VideoEdit = () => (
  <div>Video Edit</div>
);

export default VideoEdit;
