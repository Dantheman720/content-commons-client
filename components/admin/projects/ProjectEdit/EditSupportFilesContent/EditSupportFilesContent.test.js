import { mount, shallow } from 'enzyme';
import toJSON from 'enzyme-to-json';
import wait from 'waait';
import { projects } from 'components/admin/projects/ProjectEdit/mockData';
import EditSupportFilesContent from './EditSupportFilesContent';

const props = {
  data: projects[0].supportFiles.other,
  fileType: 'other',
  closeEditModal: jest.fn()
};

const Component = <EditSupportFilesContent { ...props } />;

describe( '<EditSupportFilesContent />', () => {
  it( 'renders without crashing', () => {
    const wrapper = shallow( Component );
    expect( wrapper.exists() ).toEqual( true );
    expect( toJSON( wrapper ) ).toMatchSnapshot();
  } );

  it( '`componentDidMount` sets `_isMounted`', () => {
    const wrapper = shallow( Component );
    wrapper.instance().componentDidMount();
    expect( wrapper.instance()._isMounted ).toEqual( true );
  } );

  it( '`componentWillUnmount` sets `_isMounted` to false', () => {
    const wrapper = shallow( Component );
    wrapper.instance().componentWillUnmount();
    expect( wrapper.instance()._isMounted ).toEqual( false );
  } );

  it( '`getFileExtension` returns a file\'s extension', () => {
    const wrapper = shallow( Component );
    const fileName = 'madeinamerica_arabic.srt';
    const extension = wrapper.instance().getFileExtension( fileName );
    expect( extension ).toEqual( '.srt' );
  } );

  it( '`getFileExtensions` returns an array of unique file extensions', () => {
    const wrapper = shallow( Component );
    const extensions = wrapper.instance().getFileExtensions( props.data );
    expect( extensions ).toEqual( ['.jpg', '.mp3'] );
  } );

  it( '`haveAllLangsBeenPopulated` does not set `hasPopulatedLanguages` state if all languages have not been populated', () => {
    const wrapper = shallow( Component );

    expect( wrapper.state( 'selectedLangValues' ) ).toEqual( {} );

    wrapper.setState( {
      selectedLangValues: {
        5682: 'Arabic',
        5683: 'Chinese (Simplified)',
        5684: 'English'
      }
    } );

    wrapper.instance().haveAllLangsBeenPopulated();
    expect( wrapper.state( 'selectedLangValues' ) )
      .toEqual( {
        5682: 'Arabic',
        5683: 'Chinese (Simplified)',
        5684: 'English'
      } );
    expect( wrapper.state( 'hasPopulatedLanguages' ) )
      .toEqual( false );
  } );

  it( '`haveAllLangsBeenPopulated` does set `hasPopulatedLanguages` state if all languages have been populated', () => {
    const wrapper = shallow( Component );

    expect( wrapper.state( 'selectedLangValues' ) ).toEqual( {} );

    wrapper.setState( {
      selectedLangValues: {
        5682: 'Arabic',
        5683: 'Chinese (Simplified)',
        5684: 'English',
        5685: 'French',
        5686: 'English'
      }
    } );

    wrapper.instance().haveAllLangsBeenPopulated();
    expect( wrapper.state( 'selectedLangValues' ) )
      .toEqual( {
        5682: 'Arabic',
        5683: 'Chinese (Simplified)',
        5684: 'English',
        5685: 'French',
        5686: 'English'
      } );
    expect( wrapper.state( 'hasPopulatedLanguages' ) )
      .toEqual( true );
  } );

  it( '`handleChange` sets `selectedLangValues` and `hasUnsavedData` in state', () => {
    const wrapper = shallow( Component );
    const spy = jest.spyOn(
      wrapper.instance(), 'haveAllLangsBeenPopulated'
    );
    const e = {};
    const data = { id: '5682', value: 'Arabic' };

    expect( wrapper.state( 'hasUnsavedData' ) ).toEqual( false );
    expect( wrapper.state( 'selectedLangValues' ) ).toEqual( {} );

    wrapper.instance().handleChange( e, data );

    expect( wrapper.state( 'hasUnsavedData' ) ).toEqual( true );
    expect( wrapper.state( 'selectedLangValues' ) )
      .toEqual( { 5682: 'Arabic' } );
    expect( spy ).toHaveBeenCalled();
  } );

  it( '`handleSubmit` sets `displaySaveMsg`, `hasSaved`, and `hasUnsavedData` in state', () => {
    const wrapper = shallow( Component );

    wrapper.instance().handleSubmit();

    expect( wrapper.state( 'displaySaveMsg' ) ).toEqual( true );
    expect( wrapper.state( 'hasSaved' ) ).toEqual( true );
    expect( wrapper.state( 'hasUnsavedData' ) ).toEqual( false );
  } );

  it( '`handleSubmit` calls `delayUnmount` and `handleDisplaySaveMsg` as callback', async () => {
    const wrapper = shallow( Component );
    const delayUnmountSpy = jest.spyOn( wrapper.instance(), 'delayUnmount' );
    const handleDisplaySaveMsgSpy = jest.spyOn( wrapper.instance(), 'handleDisplaySaveMsg' );

    wrapper.instance().handleSubmit();

    expect( delayUnmountSpy ).toHaveBeenCalled();
    await wait( wrapper.instance().SAVE_MSG_DELAY );
    expect( handleDisplaySaveMsgSpy ).toHaveBeenCalled();
    expect( wrapper.state( 'displaySaveMsg' ) ).toEqual( false );
    expect( wrapper.instance().saveMsgTimer ).toEqual( null );
  } );

  it( '`delayUnmount` calls a function after a delay', async () => {
    const wrapper = shallow( Component );
    const fn = jest.fn();
    const timer = () => wrapper.instance().saveMsgTimer;
    const delay = wrapper.instance().SAVE_MSG_DELAY;

    wrapper.instance().delayUnmount( fn, timer(), delay );
    await wait( delay );
    expect( fn ).toHaveBeenCalled();
  } );

  it( 'renders the `Notification` component and correct props if `displaySaveMsg` or `hasUnsavedData`', () => {
    const wrapper = shallow( Component );
    const notification = () => wrapper.find( 'Notification' );

    expect( notification().exists() ).toEqual( false );

    wrapper.setState( { displaySaveMsg: true, hasUnsavedData: false } );
    expect( notification().exists() ).toEqual( true );
    expect( notification().prop( 'el' ) ).toEqual( 'p' );
    expect( notification().prop( 'msg' ) ).toEqual( 'Saved' );
    expect( notification().prop( 'customStyles' ) )
      .toEqual( {
        position: 'absolute',
        top: '0',
        left: '50%',
        transform: 'translateX(-50%)',
        borderBottomLeftRadius: '0.28571429rem',
        borderBottomRightRadius: '0.28571429rem',
        padding: '1em 1.5em',
        fontSize: '1em'
      } );

    wrapper.setState( { displaySaveMsg: false, hasUnsavedData: false } );
    expect( notification().exists() ).toEqual( false );

    wrapper.setState( { displaySaveMsg: false, hasUnsavedData: true } );
    expect( notification().exists() ).toEqual( true );
    expect( notification().prop( 'el' ) ).toEqual( 'p' );
    expect( notification().prop( 'msg' ) )
      .toEqual( 'You have unsaved data' );
    expect( notification().prop( 'customStyles' ) )
      .toEqual( {
        position: 'absolute',
        top: '0',
        left: '50%',
        transform: 'translateX(-50%)',
        borderBottomLeftRadius: '0.28571429rem',
        borderBottomRightRadius: '0.28571429rem',
        padding: '1em 1.5em',
        fontSize: '1em'
      } );
  } );

  it( 'renders the Save button with correct message', () => {
    const wrapper = shallow( Component );
    const saveBtn = () => wrapper.find( 'Button.save' );

    expect( saveBtn().prop( 'content' ) ).toEqual( 'Save' );
    wrapper.setState( { hasSaved: true, hasUnsavedData: false } );
    expect( saveBtn().prop( 'content' ) ).toEqual( 'Data Saved' );
  } );

  it( 'renders the Save button with correct disabled state', () => {
    const wrapper = shallow( Component );
    const saveBtn = () => wrapper.find( 'Button.save' );

    expect( saveBtn().prop( 'disabled' ) ).toEqual( true );
    wrapper.setState( {
      hasSaved: true,
      hasUnsavedData: false,
      hasPopulatedLanguages: true
    } );
    expect( saveBtn().prop( 'disabled' ) ).toEqual( true );

    wrapper.setState( {
      hasSaved: false,
      hasUnsavedData: true,
      hasPopulatedLanguages: false
    } );
    expect( saveBtn().prop( 'disabled' ) ).toEqual( true );

    wrapper.setState( {
      hasSaved: false,
      hasUnsavedData: true,
      hasPopulatedLanguages: true
    } );
    expect( saveBtn().prop( 'disabled' ) ).toEqual( false );

    wrapper.setState( {
      hasSaved: true,
      hasUnsavedData: false,
      hasPopulatedLanguages: false
    } );
    expect( saveBtn().prop( 'disabled' ) ).toEqual( true );

    wrapper.setState( {
      hasSaved: true,
      hasUnsavedData: true,
      hasPopulatedLanguages: true
    } );
    expect( saveBtn().prop( 'disabled' ) ).toEqual( false );

    wrapper.setState( {
      hasSaved: false,
      hasUnsavedData: false,
      hasPopulatedLanguages: false
    } );
    expect( saveBtn().prop( 'disabled' ) ).toEqual( true );
  } );

  it( 'clicking the Save button calls the `handleSubmit`', () => {
    const wrapper = shallow( Component );
    const saveBtn = () => wrapper.find( 'Button.save' );
    const spy = jest.spyOn( wrapper.instance(), 'handleSubmit' );

    wrapper.setState( {
      hasSaved: false,
      hasUnsavedData: true,
      hasPopulatedLanguages: true
    } );
    expect( saveBtn().prop( 'disabled' ) ).toEqual( false );

    saveBtn().simulate( 'click' );
    expect( spy ).toHaveBeenCalled();
  } );

  it( 'renders the Cancel button with the correct content', () => {
    const wrapper = shallow( Component );
    const cancelCloseBtn = () => wrapper.find( 'Button.cancel-close' );

    wrapper.setState( { hasSaved: true } );
    expect( cancelCloseBtn().prop( 'content' ) ).toEqual( 'Close' );

    wrapper.setState( { hasSaved: false } );
    expect( cancelCloseBtn().prop( 'content' ) ).toEqual( 'Cancel' );
  } );

  it( '`handleCancelClose` calls `closeEditModal`', () => {
    const wrapper = shallow( Component );
    wrapper.instance().handleCancelClose();
    expect( props.closeEditModal ).toHaveBeenCalled();
  } );

  it( 'clicking the Cancel button calls `handleCancelClose`', () => {
    const wrapper = shallow( Component );
    const cancelCloseBtn = wrapper.find( 'Button.cancel-close' );
    // const spy = jest.spyOn( wrapper.instance(), 'handleCancelClose' );

    cancelCloseBtn.simulate( 'click' );
    /**
     * @todo Need to figure out why this spy assertion fails
     * even though if its return fn, props.closeEditModal, is called.
     * This would be evidence that the spy was called.
     */
    // expect( spy ).toHaveBeenCalled();
    expect( props.closeEditModal ).toHaveBeenCalled();
  } );

  it( 'renders `addFilesInputRef`', () => {
    const wrapper = mount( Component );
    const input = wrapper.find( 'input#upload-file--multiple' );
    expect( input.exists() ).toEqual( true );
    expect( wrapper.instance().addFilesInputRef ).toBeTruthy();
  } );

  it( '`null` or `[]` `data` does not crash the component', () => {
    const wrapper = shallow( Component );

    wrapper.setProps( { data: null } );
    expect( wrapper.html() ).toEqual( null );
    wrapper.setProps( { data: [] } );
    expect( wrapper.html() ).toEqual( null );
  } );
} );
