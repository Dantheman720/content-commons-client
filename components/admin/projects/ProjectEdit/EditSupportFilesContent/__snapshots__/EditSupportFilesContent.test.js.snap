// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<EditSupportFilesContent /> renders without crashing 1`] = `
<ModalItem
  className="edit-support-files other"
  headline="Edit Other files in this project"
  textDirection="ltr"
>
  <Form
    as="form"
  >
    <Table
      as="table"
      basic="very"
    >
      <TableHeader
        as="thead"
      >
        <TableRow
          as="tr"
          cellAs="td"
        >
          <TableHeaderCell
            as="th"
            width={9}
          >
            Other Files Selected
          </TableHeaderCell>
          <TableHeaderCell
            as="th"
            width={4}
          >
            Language
            <small
              className="msg--required"
            >
               *
            </small>
          </TableHeaderCell>
          <TableHeaderCell
            as="th"
          />
        </TableRow>
      </TableHeader>
      <TableBody
        as="tbody"
      >
        <EditSupportFileRow
          file={
            Object {
              "file": "madeinamerica_arabic.jpg",
              "id": "5682",
              "lang": "Arabic",
              "size": Object {
                "filesize": 58368,
              },
              "uploadStatus": Object {
                "error": false,
                "success": false,
              },
            }
          }
          fileExtensions={
            Array [
              ".jpg",
              ".mp3",
            ]
          }
          handleChange={[Function]}
          key="5682"
        />
        <EditSupportFileRow
          file={
            Object {
              "file": "madeinamerica_chinese.jpg",
              "id": "5683",
              "lang": "Chinese (Simplified)",
              "size": Object {
                "filesize": 58368,
              },
              "uploadStatus": Object {
                "error": false,
                "success": false,
              },
            }
          }
          fileExtensions={
            Array [
              ".jpg",
              ".mp3",
            ]
          }
          handleChange={[Function]}
          key="5683"
        />
        <EditSupportFileRow
          file={
            Object {
              "file": "madeinamerica_english.jpg",
              "id": "5684",
              "lang": "English",
              "size": Object {
                "filesize": 58368,
              },
              "uploadStatus": Object {
                "error": false,
                "success": false,
              },
            }
          }
          fileExtensions={
            Array [
              ".jpg",
              ".mp3",
            ]
          }
          handleChange={[Function]}
          key="5684"
        />
        <EditSupportFileRow
          file={
            Object {
              "file": "madeinamerica_french.jpg",
              "id": "5685",
              "lang": "French",
              "size": Object {
                "filesize": 58368,
              },
              "uploadStatus": Object {
                "error": false,
                "success": false,
              },
            }
          }
          fileExtensions={
            Array [
              ".jpg",
              ".mp3",
            ]
          }
          handleChange={[Function]}
          key="5685"
        />
        <EditSupportFileRow
          file={
            Object {
              "file": "madeinamerica_english.mp3",
              "id": "5686",
              "lang": "English",
              "size": Object {
                "filesize": 5242880,
              },
              "uploadStatus": Object {
                "error": false,
                "success": false,
              },
            }
          }
          fileExtensions={
            Array [
              ".jpg",
              ".mp3",
            ]
          }
          handleChange={[Function]}
          key="5686"
        />
      </TableBody>
    </Table>
    <div
      className="btn-group"
    >
      <Button
        as="button"
        basic={true}
        className="cancel-close"
        content="Cancel"
        onClick={[Function]}
        size="tiny"
        type="button"
      />
      <Button
        as="button"
        basic={true}
        className="add-files"
        color="blue"
        content="Add Files"
        onClick={[Function]}
        size="tiny"
        type="button"
      />
      <VisuallyHidden>
        <label
          htmlFor="upload-file--multiple"
        >
          upload files
        </label>
        <input
          id="upload-file--multiple"
          multiple={true}
          tabIndex={-1}
          type="file"
        />
      </VisuallyHidden>
      <Button
        as="button"
        className="save"
        color="blue"
        content="Save"
        disabled={true}
        onClick={[Function]}
        size="tiny"
        type="submit"
      />
    </div>
  </Form>
</ModalItem>
`;
